management/server/policy.go

```diff
@@ -0,0 +1,377 @@
+package server
+
+import (
+	"context"
+	_ "embed"
+	"fmt"
+	"strings"
+
+	"github.com/netbirdio/netbird/management/server/activity"
+	"github.com/netbirdio/netbird/management/server/status"
+
+	"github.com/open-policy-agent/opa/rego"
+	log "github.com/sirupsen/logrus"
+)
+
+const (
+	// UpdatePolicyName indicates a policy name update operation
+	UpdatePolicyName PolicyUpdateOperationType = iota
+	// UpdatePolicyDescription indicates a policy description update operation
+	UpdatePolicyDescription
+	// UpdatePolicyStatus indicates a policy status update operation
+	UpdatePolicyStatus
+	// UpdatePolicyQuery indicates a policy query update operation
+	UpdatePolicyQuery
+)
+
+// PolicyUpdateOperationType operation type
+type PolicyUpdateOperationType int
+
+// PolicyUpdateOperation operation object with type and values to be applied
+type PolicyUpdateOperation struct {
+	Type   PolicyUpdateOperationType
+	Values []string
+}
+
+//go:embed rego/default_policy_module.rego
+var defaultPolicyModule string
+
+//go:embed rego/default_policy.rego
+var defaultPolicy string
+
+// Policy of the Rego query
+type Policy struct {
+	// ID of the policy
+	ID string
+
+	// Name of the Policy
+	Name string
+
+	// Description of the policy visible in the UI
+	Description string
+
+	// Disabled status of the policy
+	Disabled bool
+
+	// Query of Rego the policy
+	Query string
+}
+
+// Copy returns a copy of the policy.
+func (r *Policy) Copy() *Policy {
+	return &Policy{
+		ID:          r.ID,
+		Name:        r.Name,
+		Description: r.Description,
+		Disabled:    r.Disabled,
+		Query:       r.Query,
+	}
+}
+
+// EventMeta returns activity event meta related to this policy
+func (p *Policy) EventMeta() map[string]any {
+	return map[string]any{"name": p.Name}
+}
+
+// FirewallRule is a rule of the firewall.
+type FirewallRule struct {
+	// PeerID of the peer
+	PeerID string
+
+	// PeerIP of the peer
+	PeerIP string
+
+	// Direction of the traffic
+	Direction string
+
+	// Action of the traffic
+	Action string
+
+	// Port of the traffic
+	Port string
+}
+
+// parseFromRegoResult parses the Rego result to a FirewallRule.
+func (f *FirewallRule) parseFromRegoResult(value interface{}) error {
```

Comment:
With the struct tags and reflect package feature would not be nicer the parse?
```
type FirewallRule struct {
  PeerId string `rego:"ID"`
  ...
}
```

Reply:
Thanks for the note, reflect is not a good choice when you do simple things like this. I build reflect parsers only when I need to parse different types of structs in the same way, in that case, it's ok, but here it will be overkill.

Reply:
Thank you your answer. You are right.